libdevcore/MemoryDB.cpp

```diff
@@ -14,176 +14,95 @@
     You should have received a copy of the GNU General Public License
     along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
-/** @file MemoryDB.cpp
- * @author Gav Wood <i@gavwood.com>
- * @date 2014
- */
 
 #include "MemoryDB.h"
-#include "Common.h"
-#include "CommonData.h"
-using namespace std;
-using namespace dev;
 
 namespace dev
 {
-
-std::unordered_map<h256, std::string> MemoryDB::get() const
+namespace db
 {
-#if DEV_GUARDED_DB
-    ReadGuard l(x_this);
-#endif
-    std::unordered_map<h256, std::string> ret;
-    for (auto const& i: m_main)
-        if (!m_enforceRefs || i.second.second > 0)
-            ret.insert(make_pair(i.first, i.second.first));
-    return ret;
-}
 
-MemoryDB& MemoryDB::operator=(MemoryDB const& _c)
+void MemoryDBWriteBatch::insert(Slice _key, Slice _value)
 {
-    if (this == &_c)
-        return *this;
-#if DEV_GUARDED_DB
-    ReadGuard l(_c.x_this);
-    WriteGuard l2(x_this);
-#endif
-    m_main = _c.m_main;
-    m_aux = _c.m_aux;
-    return *this;
-}
-
-std::string MemoryDB::lookup(h256 const& _h) const
-{
-#if DEV_GUARDED_DB
-    ReadGuard l(x_this);
-#endif
-    auto it = m_main.find(_h);
-    if (it != m_main.end())
-    {
-        if (!m_enforceRefs || it->second.second > 0)
-            return it->second.first;
-        else
-            cwarn << "Lookup required for value with refcount == 0. This is probably a critical trie issue" << _h;
-    }
-    return std::string();
+    m_batch[_key.toString()] = _value.toString();
 }
 
-bool MemoryDB::exists(h256 const& _h) const
+void MemoryDBWriteBatch::kill(Slice _key)
 {
-#if DEV_GUARDED_DB
-    ReadGuard l(x_this);
-#endif
-    auto it = m_main.find(_h);
-    if (it != m_main.end() && (!m_enforceRefs || it->second.second > 0))
-        return true;
-    return false;
+    m_batch.erase(_key.toString());
 }
 
-void MemoryDB::insert(h256 const& _h, bytesConstRef _v)
+std::string MemoryDB::lookup(Slice _key) const
 {
-#if DEV_GUARDED_DB
-    WriteGuard l(x_this);
-#endif
-    auto it = m_main.find(_h);
-    if (it != m_main.end())
+    std::string value;
+    Guard lock(m_mutex);
+    auto const& it = m_db.find(_key.toString());
```

Comment:
This can be simplified a bit as
```
if (it != m_db.end())
    return it->second;
return {};
```

Reply:
Also, for single-liners after `if` we skip `{}`.

Reply:
Thanks! Was aware of the skipping {} for single-line ifs but forgot in this case, thanks for catching this!